### DISCUSSION TITLE
Setting OAuth2PasswordBearer tokenUrl with root-path

### DISCUSSION THREAD

**Discussion by imacat** [fastapi/fastapi#6045] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import datetime, timedelta
from typing import Union

from fastapi import Depends, FastAPI, HTTPException, status
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from jose import JWTError, jwt
from passlib.context import CryptContext
from pydantic import BaseModel

# to get a string like this run:
# openssl rand -hex 32
SECRET_KEY = "09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7"
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 30


fake_users_db = {
    "johndoe": {
        "username": "johndoe",
        "full_name": "John Doe",
        "email": "johndoe@example.com",
        "hashed_password": "$2b$12$EixZaYVK1fsbw1ZfbX3OXePaWxn96p36WQoeG6Lruj3vjPGga31lW",
        "disabled": False,
    }
}


class Token(BaseModel):
    access_token: str
    token_type: str


class TokenData(BaseModel):
    username: Union[str, None] = None


class User(BaseModel):
    username: str
    email: Union[str, None] = None
    full_name: Union[str, None] = None
    disabled: Union[bool, None] = None


class UserInDB(User):
    hashed_password: str


pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

app = FastAPI()


def verify_password(plain_password, hashed_password):
    return pwd_context.verify(plain_password, hashed_password)


def get_password_hash(password):
    return pwd_context.hash(password)


def get_user(db, username: str):
    if username in db:
        user_dict = db[username]
        return UserInDB(**user_dict)


def authenticate_user(fake_db, username: str, password: str):
    user = get_user(fake_db, username)
    if not user:
        return False
    if not verify_password(password, user.hashed_password):
        return False
    return user


def create_access_token(data: dict, expires_delta: Union[timedelta, None] = None):
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=15)
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)
    return encoded_jwt


async def get_current_user(token: str = Depends(oauth2_scheme)):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exception
        token_data = TokenData(username=username)
    except JWTError:
        raise credentials_exception
    user = get_user(fake_users_db, username=token_data.username)
    if user is None:
        raise credentials_exception
    return user


async def get_current_active_user(current_user: User = Depends(get_current_user)):
    if current_user.disabled:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user


@app.post("/token", response_model=Token)
async def login_for_access_token(form_data: OAuth2PasswordRequestForm = Depends()):
    user = authenticate_user(fake_users_db, form_data.username, form_data.password)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
            headers={"WWW-Authenticate": "Bearer"},
        )
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": user.username}, expires_delta=access_token_expires
    )
    return {"access_token": access_token, "token_type": "bearer"}


@app.get("/users/me/", response_model=User)
async def read_users_me(current_user: User = Depends(get_current_active_user)):
    return current_user


@app.get("/users/me/items/")
async def read_own_items(current_user: User = Depends(get_current_active_user)):
    return [{"item_id": "Foo", "owner": current_user.username}]
```

### Description

Hi.  I'm creating a protected API service that is mounted at `/api` with `uvicorn --root-path /api`.  It is protected with `OAuth2PasswordBearer`, as described in [FastAPI Security Tutorial](https://fastapi.tiangolo.com/tutorial/security/).  However, `tokenUrl` in `openapi.json` does not respect `root_path`.  When I'm working on my local development server without `--root-path /api`, `tokenUrl` in `openapi.json` is `/token`.  When it is deployed onto my server with `--root-path /api`, it is still `/token`, where it should be `/api/token`.  As a result, the Swagger UI log in always fails.

Currently I did a dirty fix in the `custom_openapi` function:

    def custom_openapi()
        if app.openapi_schema:
            return app.openapi_schema
        openapi_schema = get_openapi()
        if "servers" in openapi_schema:
            root_path: str = openapi_schema["servers"][0]["url"]
            openapi_schema["components"]["securitySchemes"][
                "OAuth2PasswordBearer"]["flows"]["password"]["tokenUrl"] \
                = root_path + openapi_schema["components"]["securitySchemes"][
                "OAuth2PasswordBearer"]["flows"]["password"]["tokenUrl"]
        app.openapi_schema = openapi_schema
        return app.openapi_schema
    
    app: FastAPI = FastAPI()
    app.openapi = custom_openapi

But this does not work if I have multiple servers in the future.

How can I make `tokenUrl` to respect the current `root_path`?  Did I missed something?

### Operating System

Linux

### FastAPI Version

0.88.0

### Python Version

3.9.2

**Comment by iudeen** [comment]
Did you try with 
```python
app = FastAPI(root_path = "/api")
```

**Comment by imacat** [comment]
@iudeen Thank you.  This seems to be a workaround to find the current `root_path`, although I still have to figure out what to pass as the `root_path` from uvicorn.

And it still does not solve the problem that `tokenUrl` does not respect `root_path`.

**Comment by Lenclove** [comment]
@imacat I think you can get some ideas from this link -- [Click Here](https://fastapi.tiangolo.com/advanced/sub-applications/). This page may help you understanding the usage of the `root-path`. The question you mentioned could be solved by the `mount` and `sub-applications`.

**Reply by abhishek-compro** [reply]
The problem is with `OAuth2PasswordBearer(tokenUrl="token")` as the `tokenUrl` is left as it is and isn't prefixed by the `root_path`. I cannot find anything that exposes the value of `root_path` so that I can prefix it by myself in the code. Either we have to follow @imacat 's hack or use an environment variable mentioned at https://github.com/tiangolo/fastapi/issues/4328#issue-1088332014 . If you know of any other simpler way which doesn't involve hacks or environment variables then please explain.

**Comment by agn-7** [comment]
I had the same problem, but my situation may be different. Actually, I am using a `router` with the prefix `/api`, so with `oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")` I got `404 not found`. Therefore, to solve the issue I had to use `"api/token"`:

`oauth2_scheme = OAuth2PasswordBearer(tokenUrl="api/token")`

**Comment by YuriiMotov** [comment, marked as answer]
Why do you need to add this prefix?

It was probably different in FastAPI 0.88, but now everything works correctly.

`tokenUrl` is still just `token` in openapi:
```
   "securitySchemes": {
      "OAuth2PasswordBearer": {
        "type": "oauth2",
        "flows": {
          "password": {
            "scopes": {

            },
            "tokenUrl": "token"
          }
        }
      }
    }
```

But combining it with value from servers:
```
  "servers": [
    {
      "url": "/api"
    }
  ],
```
You will get correct URL.

I just checked, Swagger works well behind the Traefik with stripped `/api` prefix and app running via `uvicorn main:app --root-path /api`
